src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php

```diff
@@ -258,34 +258,13 @@ private function compileRoute(Route $route, $name, $supportsRedirections, $paren
 EOF;
 
         $gotoname = 'not_'.preg_replace('/[^A-Za-z0-9_]/', '', $name);
-        if ($methods) {
-            if (1 === count($methods)) {
-                $code .= <<<EOF
-            if (\$this->context->getMethod() != '$methods[0]') {
-                \$allow[] = '$methods[0]';
-                goto $gotoname;
-            }
-
-
-EOF;
-            } else {
-                $methods = implode("', '", $methods);
-                $code .= <<<EOF
-            if (!in_array(\$this->context->getMethod(), array('$methods'))) {
-                \$allow = array_merge(\$allow, array('$methods'));
-                goto $gotoname;
-            }
-
-
-EOF;
-            }
-        }
 
         if ($hasTrailingSlash) {
             $code .= <<<EOF
             if ('/' === substr(\$pathinfo, -1)) {
                 // no-op
             } elseif (!in_array(\$this->context->getMethod(), array('HEAD', 'GET'))) {
+                \$allow[] = 'GET';
```

Comment:
How do we know that GET is actually allowed? If do POST /foo on a PUT /foo/ route, it will say method now allowed but GET is allowed? We don't know if GET is allowed at all.

Reply:
Yes it is: the redirect handles the GET

Reply:
I guess you need to check if get is in the allowed methods of the route first.

Reply:
and if not we can mark the request with 307 response in master and here we just continue matching

Reply:
GET *is* allowed: it triggers a redirect. There is nothing else to consider on this very path without slash. The verbs allowed on the route apply only when there is a slash.

Reply:
Can you elaborate?

Reply:
I'm pretty sure there is a problem. `$supportsTrailingSlash = $supportsRedirections && (!$methods || in_array('HEAD', $methods));`
Again to my example. Let's say the only route is PUT/HEAD /foo/
A request to POST /foo would say method not allowed and GET is allowed. But that is not true.

Reply:
So `$supportsTrailingSlash` should check for GET and not HEAD

Reply:
>  A request to POST /foo would say method not allowed and GET is allowed. But that is not true.

You're focusing too much on the route definition itself. Here, we don't match the route definition, but we match a behavior that the router has, standalone: trailing slash redirections.
Let's agree that there is a bug: a POST should not trigger a redirect, which is does for now.
If we want to fix this, we can throw a NotMatched.
BUT: if this were a GET, it would work actually, as the router would issue a 301/302.
Thus, the current 404 is too restrictive, and the router should hint that a GET would work better: 405.

Reply:
405 means, that the method for the current URI is not allowed but a different method for the same URI is allowed. In the RedirectableUrlMatcher you are now just returning 405 without knowing if GET to the same URI is allowed. So a client would receive a 405 and then retry the same request with GET and he could still receive a 404. Then there is no point in returning 405 in the first place.

Reply:
> retry the same request with GET and he could still receive a 404

Why? to my understanding, it would then get a 301/302, because of this very if which will be hit again.

Reply:
@Tobion note that if you're fine with the 404, but not with the 405, I'm OK to remove this line. It's only a minor thing on the HTTP side, and I doubt in will make any real difference in practice.

Reply:
ping @Tobion let's agree on this and move on :)

Reply:
If you HEAD with GET in `$supportsTrailingSlash = $supportsRedirections && (!$methods || in_array('HEAD', $methods));` this works for PhpMatcherDumper. But the logic RedirectableUrlMatcher is broken. See below.

Reply:
done